eoscompanion/content_adjusters.py

```diff
@@ -18,46 +18,283 @@
-    be resolved by the server.
+def maybe_ekn_id_to_server_uri(ekn_uri, query):
+    '''Convert an EKN URI (ekn:///id) to a URI that resolves on this server.
 
-    Right now the way that this is done is a total hack (regex), in future
-    we might want to depend on beautifulsoup and use that instead.
+    If :ekn_uri: is None, return None.
     '''
-    return EosCompanionAppService.string_to_bytes(
-        _RE_EKN_URL_CAPTURE.sub(ekn_url_rewriter(query),
-                                EosCompanionAppService.bytes_to_string(content_bytes))
-    )
+    if ekn_uri is None:
+        return None
+
+    return rewrite_ekn_url(os.path.basename(urlparse(ekn_uri).path), query)
+
+
+_MOBILE_WRAPPER_TEMPLATE_URI = (
+    'resource:///com/endlessm/CompanionAppService/data/templates/mobile-article-wrapper.mst'
+)
+
+
+def render_mobile_wrapper(renderer,
+                          app_id,
+                          rendered_content,
+                          metadata,
+                          engine,
+                          shards,
+                          query,
+                          callback):
+    '''Render the page wrapper and initialize crosslinks.
+
+    This is the final rendering step that EKN content needs to undergo before
+    it becomes usable, at least on the desktop. This step initializes
+    both a dictionary of metadata about the content and also a map of
+    outgoing links to internal links. Then we inject some javascript which
+    at browser-render time, rewrites the page to resolve all those links.
+    '''
+    def _on_queried_sets(src, result):
+        '''Called when we finish querying set objects.'''
+        try:
+            results = src.query_finish(result)
+        except GLib.Error as error:
+            callback(error, None)
+            return
+
+        set_objects = results.get_property('models')
+        sets_for_tags = {
+            tag: set_object
+            for tag, set_object in tagged_sets(set_objects)
+            if not tag.startswith('Ekn')
+        }
+        content_metadata = {
+            'title': metadata.get('title', ''),
+            'published': metadata.get('published', ''),
+            'authors': metadata.get('authors', []),
+            'license': metadata.get('license', ''),
+            'source': metadata.get('source', ''),
+            'source_name': metadata.get('sourceName', ''),
+            'originalURI': metadata.get('originalURI', ''),
+            'sets': [
+                {
+                    'child_tags': set_object.get_property('child-tags'),
+                    'ekn_id': set_object.get_property('ekn-id'),
+                    'title': set_object.get_property('title'),
+                    'tags': set_object.get_property('tags')
+                }
+                for set_object in
+                yield_set_objects_for_tags(sets_for_tags,
+                                           metadata.get('tags', []))
```

Comment:
Would it be possible to make the generation of the `set_object` list a bit simpler? Right now we take `set_objects` above, we make another dictionary by flattening them, but then in `yield_set_objects_for_tags()` we kind of loop over it again to use another list of tags coming from the metadata. I wonder if we could not already consider the metadata tags when building the first dictionary.

Reply:
Possibly. This was basically just a straight translation of code from EKL, so I'll look into it.